### ISSUE TITLE
Linter plugins: Support plugins with `/` in name

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#14557]
See #14503 for original bug report.

`.oxlintrc.json`:

```json
{
	"jsPlugins": ["@eslint-react/eslint-plugin"],
	"rules": {
		"@eslint-react/eslint-plugin/jsx-no-comment-textnodes": ["error"],
		"@eslint-react/eslint-plugin/dom/no-dangerously-set-innerhtml": ["error"],
		"@eslint-react/eslint-plugin/naming-convention/context-name": ["error"]
	}
}
```

Result:

```
Failed to parse configuration file.
  x Plugin '@eslint-react' not found
```

### The problem

I assume the problem is that we split complete rule name into plugin name and rule name on the first `/`. But in this case, the plugin name contains a slash.

Splitting on the last slash wouldn't work either, as rule names also contain a `/`.

### Possible solutions

1. For each key in `rules`, try all possible combinations - i.e.
  * Plugin `@eslint-react`, rule `eslint-plugin/naming-convention/context-name`
  * Plugin `@eslint-react/eslint-plugin`, rule `naming-convention/context-name`
  * Plugin `@eslint-react/eslint-plugin/naming-convention`, rule `context-name`

2. Automatically remove `eslint-plugin-` prefix or `/eslint-plugin` postfix from plugin names.

Config would then look like:

```json
{
	"jsPlugins": ["@eslint-react/eslint-plugin"],
	"rules": {
		"@eslint-react/jsx-no-comment-textnodes": ["error"],
		"@eslint-react/dom/no-dangerously-set-innerhtml": ["error"],
		"@eslint-react/naming-convention/context-name": ["error"]
	}
}
```

We should probably look at what ESLint does.

**Comment by Rel1cx** [comment]
https://github.com/johnsoncodehk/tsslint/issues/41 TSSLint had the same issue before and they solved it. It seems we can refer to their solution https://github.com/johnsoncodehk/tsslint/commit/9b1e0f47698ef560bd35b927e0b3cc5faafb6eee.

**Comment by camc314** [comment]
cc @overlookmotel we should just copy this exactly i think. we don't need to try the alternative combinations

https://github.com/eslint/eslint/blob/b0674be94e4394401b4f668453a473572c321023/lib/config/config.js#L134-L156

**Comment by overlookmotel** [comment]
@Rel1cx I see you're the maintainer of `@eslint-react/eslint-plugin`. Can I ask for your help?

Please excuse my stupidity, but I'm thoroughly confused by this. If a user of `@eslint-react/eslint-plugin` wants to enable a rule from one of the inherited packages, what do they put in ESLint config? Something like this?

```js
rules: {
  "@eslint-react/naming-convention/context-name": "error",
}
```

If so, how does that work with ESLint, since it appears to split plugin name from rule name on last `/` when ID starts with `@`:

```js
if (ruleId.includes("/")) {
	// mimic scoped npm packages
	if (ruleId.startsWith("@")) {
		pluginName = ruleId.slice(0, ruleId.lastIndexOf("/"));
	} else {
		pluginName = ruleId.slice(0, ruleId.indexOf("/"));
	}

	ruleName = ruleId.slice(pluginName.length + 1);
}
```

[(source)](https://github.com/eslint/eslint/blob/b0674be94e4394401b4f668453a473572c321023/lib/config/config.js#L134-L156)

And the name of the plugin isn't `@eslint-react/naming-convention`?

**Comment by Rel1cx** [comment]
@overlookmotel Thanks for reaching out, and that's an good question. The confusion is completely understandable because of how ESLint's configuration and rule parsing have evolved.

Let me address your points one by one:

> If a user of `@eslint-react/eslint-plugin` wants to enable a rule from one of the inherited packages, what do they put in ESLint config? Something like this?
>
> ```js
> rules: {
>   "@eslint-react/naming-convention/context-name": "error",
> }
> ```

Yes, that is exactly right for the modern **flat config**.

> If so, how does that work with ESLint, since it appears to split plugin name from rule name on last `/` when ID starts with `@`
> ...
> And the name of the plugin isn't `@eslint-react/naming-convention`?

You've hit on the core of the issue. With the introduction of the **flat config**, ESLint's parsing logic changed.

*   **Legacy Config (e.g., `.eslintrc`)**: The plugin was `@eslint-react`, and the rule name was `naming-convention/context-name`. The second `/` was just a character in the rule name (it’s only meaningful semantically, and doesn’t affect parsing).
*   **Flat Config (e.g., `eslint.config.js`)**: As you pointed out, ESLint now parses `@eslint-react/naming-convention` as the plugin name and `context-name` as the rule name.

To maintain compatibility with both systems, `@eslint-react/eslint-plugin` registers `@eslint-react/naming-convention` as a "sub-plugin" when used in a flat config environment. This makes your configuration work as expected without requiring users to change rule names during migration.

So, in essence, for flat config, `@eslint-react/naming-convention` *is* treated as the plugin name, just as your investigation into ESLint's source code suggests.

**Comment by Rel1cx** [comment]
So, the config file as shown below is closer to ESLint’s current behavior:

```json
{
  "jsPlugins": [
    "@eslint-react",
    "@eslint-react/dom",
    "@eslint-react/naming-convention"
  ],
  "rules": {
    "@eslint-react/jsx-no-comment-textnodes": ["error"],
    "@eslint-react/dom/no-dangerously-set-innerhtml": ["error"],
    "@eslint-react/naming-convention/context-name": ["error"]
  }
}
```

Or just use the [unscoped plugins](https://github.com/Rel1cx/eslint-react?tab=readme-ov-file#modular) from ESLint React:

```json
{
  "jsPlugins": [
    "react-x",
    "react-dom",
    "react-naming-convention"
  ],
  "rules": {
    "react-x/jsx-no-comment-textnodes": ["error"],
    "react-dom/no-dangerously-set-innerhtml": ["error"],
    "react-naming-convention/context-name": ["error"]
  }
}
```

cc @magic-akari

**Comment by Rel1cx** [comment]
[@overlookmotel](https://github.com/overlookmotel) I have reorganized the reply in https://github.com/oxc-project/oxc/issues/14557#issuecomment-3399160211, and I hope it can better answer your questions this time.

**Comment by BeyramTaglietti** [comment]
Is there any solution currently to this problem?

**Comment by camc314** [comment]
Not yet

**Comment by camc314** [comment]
This is also a problem for eslint plugin tanstack query:

```json
{
    "jsPlugins": ["@tanstack/eslint-plugin-query"],
    "categories": {
        "correctness": "off"
    },
    "rules": {
        "@tanstack/query/exhaustive-deps": "error",
        "@tanstack/query/no-deprecated-options": "error",
        "@tanstack/query/prefer-query-object-syntax": "error",
        "@tanstack/query/stable-query-client": "error"
    }
}
```

Error:

```
./node_modules/.bin/oxlint
WARNING: JS plugins are experimental and not subject to semver.
Breaking changes are possible while JS plugins support is under development.
Failed to parse configuration file.

  × Plugin '@tanstack' not found
```

**Comment by overlookmotel** [comment]
We now normalize plugin names same as ESLint does. To take the Tanstack example:

```json
{
    "jsPlugins": ["@tanstack/eslint-plugin-query"],
    "rules": {
        "@tanstack/query/exhaustive-deps": "error",
    }
}
```

The plugin name gets set to `@tanstack/query` in `ExternalPlugin`, and `context.id` on JS side is `@tanstack/query/exhaustive-deps` (once #17033 is merged).

But we currently get errors because when parsing the rule name, it splits it on 1st `/`, not the last. So it divides into:

* Plugin name: `@tanstack`
* Rule name: `query/exhaustive-deps`
 
It should be:

* Plugin name: `@tanstack/query`
* Rule name: `exhaustive-deps`

So I think we can fix this by:

1. When splitting rule names in config, split on last `/` instead of first.
2. Ditto in `// oxlint-disable @tanstack/query/exhaustive-deps` comments (unless that's what we already do).

@camc314 Do you agree? Is it as simple as that?

**Comment by overlookmotel** [comment]
I'm going to assume this is the way to solve it.

@copilot please action my comment above. Make your PR on top of #17033.

**Comment by overlookmotel** [comment]
Actually, no the answer is this: https://github.com/oxc-project/oxc/issues/14557#issuecomment-3397729751

```js
if (ruleId.includes("/")) {
	// mimic scoped npm packages
	if (ruleId.startsWith("@")) {
		pluginName = ruleId.slice(0, ruleId.lastIndexOf("/"));
	} else {
		pluginName = ruleId.slice(0, ruleId.indexOf("/"));
	}

	ruleName = ruleId.slice(pluginName.length + 1);
}
```

**Comment by overlookmotel** [comment]
Handling of rule names containing multiple slashes in config is now fixed (#17073).

What remains is to correctly handle them in comments e.g.:

```js
// oxlint-disable @tanstack/query/exhaustive-deps
```

I set Claude on it, but it looks more complex than I expected, and I'm not familiar with that logic, so I'm leaving it for now.

Workaround in meantime is to define an alias in config, so the plugin name doesn't contain slashes any more: https://oxc.rs/docs/guide/usage/linter/js-plugins.html#plugin-aliases

**Comment by connorshea** [comment]
I think it's probably worth splitting the last bit of work (disable directives) into its own ticket?

**Comment by connorshea** [comment]
I am actually running into problems with this, where I'm not sure how to use these nested rules?

If I have a config like this (see https://github.com/oxc-project/oxlint-migrate/pull/402 and https://github.com/oxc-project/oxlint-migrate/issues/377):
```js
import js from '@eslint/js'
import globals from 'globals'
import reactHooks from 'eslint-plugin-react-hooks'
import reactRefresh from 'eslint-plugin-react-refresh'
import { defineConfig, globalIgnores } from 'eslint/config'
import eslintReact from "@eslint-react/eslint-plugin"

export default defineConfig([
  globalIgnores(['dist']),
  {
    files: ['**/*.{js,jsx}'],
    extends: [
      js.configs.recommended,
      reactHooks.configs.flat.recommended,
      reactRefresh.configs.vite,
      eslintReact.configs.recommended,
    ],
    languageOptions: {
      ecmaVersion: 2020,
      globals: globals.browser,
      parserOptions: {
        ecmaVersion: 'latest',
        ecmaFeatures: { jsx: true },
        sourceType: 'module',
      },
    },
    rules: {
      'no-unused-vars': ['error', { varsIgnorePattern: '^[A-Z_]' }],
    },
  },
])
```

How am I supposed to use the rules from `eslintReact.configs.recommended`? And how am I supposed to set the ESLint plugin names? 

The meta.name for the plugin is set to the package name here: https://github.com/Rel1cx/eslint-react/blob/v2.13.0/packages/plugins/eslint-plugin-react-dom/src/plugin.ts#L3

Which means the meta name is `eslint-plugin-react-dom`, and even if I set the `jsPlugins` value to include `eslint-plugin-react-dom`, I get errors about the rule names here.

Config:

```jsonc
{
  "rules": {
    "@eslint-react/dom/no-find-dom-node": "error",
    "@eslint-react/dom/no-flush-sync": "error",
    "@eslint-react/dom/no-hydrate": "error",
    "@eslint-react/dom/no-namespace": "error",
    "@eslint-react/dom/no-render": "error",
    "@eslint-react/dom/no-render-return-value": "error",
    "@eslint-react/dom/no-script-url": "warn",
    "@eslint-react/dom/no-unsafe-iframe-sandbox": "warn",
    "@eslint-react/dom/no-use-form-state": "error",
    "@eslint-react/dom/no-void-elements-with-children": "error",
  },
  "jsPlugins": [
    "@eslint-react/eslint-plugin",
    "eslint-plugin-react-dom",
  ]
}
```

Result:

```
Failed to build configuration from /Users/connorshea/code/eslint-react-oxc-migration/.oxlintrc.json.

  × Plugin '@eslint-react/dom' not found
  │ 
  │ Plugin '@eslint-react/dom' not found
  │ 
  │ Plugin '@eslint-react/dom' not found
  │ 
```

Should the rule names for nested rules be the same as the jsPlugin name? and/or is there something really weird happening with the `@eslint-react` plugin that causes problems here?

**Comment by connorshea** [comment]
I guess because of [this](https://github.com/Rel1cx/eslint-react/blob/f4e271cc717474666d492c0fe06a5c54479c3933/packages/plugins/eslint-plugin/src/configs/recommended.ts), it exports the plugins object, which you then use via `...reactPlugin.recommended` and that spreads the plugin aliases into the config?

```js
export const plugins = {
  ...x.plugins,
  ...rsc.plugins,
  ...dom.plugins,
  ...webApi.plugins,
  "@eslint-react/naming-convention": reactNamingConvention,
};
```

And so right now we don't handle that correctly in the oxlint-migrate tool, but the correct solution is probably to either:

1. Migrate these explicitly as JS Plugin aliases (to rename the plugin to the fancy `@eslint-react/naming-convention` name), or
2. Migrate it such that you add `eslint-plugin-react-naming-convention` to jsPlugins and rename `@eslint-react/naming-convention/ref-name` in the rules list to `react-naming-convention/ref-name`

The problem is that I'm unsure how to detect this behavior programmatically, short of hardcoding it.

**Comment by connorshea** [comment]
I think I've got it fixed in https://github.com/oxc-project/oxlint-migrate/pull/402 now, I went with option 1 from my last comment.

However, I'm unsure if there's anything we can/should fix to make sure this works properly somehow. I'm thoroughly confused as to what/how the ESLint plugin supports this in the first place and why our porting of it doesn't work with `@eslint-react/dom/rule-name` (excepting the usage of JS plugin aliases which seem like a hacky fix for the problem).

----

Basically, your options for using the nested rules from `@eslint-react/eslint-plugin` in Oxlint right now are either this:

```json
{
  "rules": {
    "@eslint-react/react-dom/no-find-dom-node": "error",
    "@eslint-react/react-dom/no-flush-sync": "error",
    "@eslint-react/react-dom/no-hydrate": "error",
    "@eslint-react/react-dom/no-namespace": "error",
    "@eslint-react/react-dom/no-render": "error",
    "@eslint-react/react-dom/no-render-return-value": "error",
    "@eslint-react/react-dom/no-script-url": "warn",
    "@eslint-react/react-dom/no-unsafe-iframe-sandbox": "warn",
    "@eslint-react/react-dom/no-use-form-state": "error",
    "@eslint-react/react-dom/no-void-elements-with-children": "error"
  },
  "jsPlugins": [
    { "name": "@eslint-react/dom", "specifier": "eslint-plugin-react-dom" }
  ]
}
```

Or this:

```json
{
  "rules": {
    "react-dom/no-find-dom-node": "error",
    "react-dom/no-flush-sync": "error",
    "react-dom/no-hydrate": "error",
    "react-dom/no-namespace": "error",
    "react-dom/no-render": "error",
    "react-dom/no-render-return-value": "error",
    "react-dom/no-script-url": "warn",
    "react-dom/no-unsafe-iframe-sandbox": "warn",
    "react-dom/no-use-form-state": "error",
    "react-dom/no-void-elements-with-children": "error"
  },
  "jsPlugins": [
    "eslint-plugin-react-dom"
  ]
}
```

As best as I've been able to tell, no other options work.
